### PR TITLE
feat: add Gemma3 VLM OpenXLA path (#869)

### LINKED ISSUE #869: feat(xla): support Gemma3 VLM additive prefill masks
## Parent and dependency

- Parent epic: #566
- Blocked by #863
- Existing MLX implementation: `src/loading/vlm_gemma.rs`, `src/vision/merge.rs::prepare_inputs_for_multimodal`, and `src/models/gemma3.rs`

## Problem

Gemma3 VLM uses the reusable SigLIP/projector shape but not LLaVA's ordinary causal merge. The current MLX path scales projected image features by `1 / sqrt(hidden_size)`, merges them with text embeddings, then the Gemma3 model scales the complete hidden sequence by `sqrt(hidden_size)`. It also supplies an additive f32 4D prefill mask (`0.0` for allowed valid-token pairs, `f32::MIN` for padded pairs) to every language layer instead of constructing the normal causal/sliding masks.

If the XLA embeddings entry treats the prepared values as post-scale but the preprocessor exports the pre-scale MLX merge result, text/image magnitudes diverge. If it substitutes a causal mask, the model architecture diverges independently of the vision encoder.

## Goal

Support a pinned Gemma3 VLM checkpoint end to end on XLA with exact embedding scaling, placeholder/padding semantics, additive 4D prefill attention, SigLIP/projector execution, and ordinary one-token decode.

## Required implementation

- Extend the host preprocessor with the Gemma3 prompt wrapping and image-token expansion used by `vlm_prompt.rs`/`vlm_gemma.rs`, including configured BOI/image/newline token IDs and exact image-feature cardinality.
- Reuse the IREE SigLIP/projector primitives from #863 with Gemma3 checkpoint prefixes, shapes, normalization epsilon, image size, and projector mapping validated explicitly.
- Reproduce the net embedding scale at the canonical #858 boundary: the tensor handed to `prefill_embeddings.main` must be post-Gemma embedding scale. Document and test that text positions receive `sqrt(hidden_size)` while the pre-divide/overall-scale combination leaves projected image features at the reference magnitude.
- Construct the exact additive f32 mask currently produced by `prepare_inputs_for_multimodal`, including padding rows/columns and canonical shape `[1, 1, Lp, Lp]` for one request. Do not replace it with a boolean or multiplicative mask.
- Pass the external mask to every Gemma3 prefill layer as the existing MLX path does. Decode remains the normal one-token no-mask path over the seeded KV cache.
- Validate, with an independent reference implementation, whether all valid prefill token pairs are bidirectional for the selected checkpoint/config. If upstream semantics differ by variant, encode the variant in configuration rather than guessing from image-token positions.
- Include mask mode, image token IDs, scaling policy, SigLIP/projector config, and Gemma sliding/global attention config in artifact compatibility identity.
- Integrate filtered loading, XLA capabilities, CLI/server batching, cancellation, and no-image text fallback without a full duplicate MLX decoder.

## Non-goals

- Gemma3n MobileNet/per-layer inputs or audio.
- Gemma4 unified block masks.
- General arbitrary 4D masks from untrusted API callers.

## Validation

- Pin a Gemma3 VLM checkpoint/revision and compare prompt tokens, image-token positions, raw/projected/scaled image features, text and image embedding norms, every mask cell for a small fixture, selected layer outputs, logits/KV, and greedy tokens against MLX and an independent oracle.
- Cover no padding, left/right padding as supported, one and multiple images, placeholder mismatch, all-valid mask, padded mask, no-image text fallback, and context overflow.
- Add negative fixtures for accidental causal masking, `0/1` multiplicative masking, double Gemma scaling, and missing image pre-scale; each must be caught at an intermediate assertion.
- Run CLI and mixed text/image continuous-batch server tests on a production-relevant IREE target.
- Run #862/#863 and all text-only Gemma3 sliding/global attention regression gates.

## Acceptance criteria

- [ ] Exported merged embeddings have the exact post-scale magnitudes required by #858.
- [ ] The f32 additive 4D mask matches MLX and the independent checkpoint oracle cell-for-cell.
- [ ] External mask routing reaches every prefill layer and does not affect one-token decode.
- [ ] SigLIP/projector and final deterministic output meet the documented tolerances/token exactness.
- [ ] Padding, multi-image, and no-image behavior are tested through CLI and continuous batching.
- [ ] No duplicate full MLX decoder or silent causal-mask fallback remains.
- [ ] Existing text-only Gemma3 XLA behavior remains token-exact.

### PR DESCRIPTION
**Pull request by inureyes**
## Summary

- Add the pinned Gemma3 SigLIP tower and average-pool projector to the resident StableHLO/IREE vision module, including exact processor, projector, weight-shape, and artifact-identity contracts.
- Apply the SigLIP `post_layernorm` after the final encoder block on the IREE path, matching the eager MLX tower, and version the vision artifact identity for the corrected graph.
- Add the Gemma3 prepared-prefill seam with one-time text embedding scaling, native projector rows, and the authoritative bidirectional padding mask used by embeddings prefill.
- Bind the prepared-prefill ABI, additive-mask mode/value, post-scale policy, newline/image/pad/BOI/EOI tokens, SigLIP/projector configuration, and Gemma sliding/global settings into an explicit stable identity used by both single-sequence and ragged IREE bundle fingerprints. Preserve the existing dense identity byte-for-byte for ordinary text-only Gemma3 configs.
- Parse the nested Gemma3 text configuration for the shared language emitter and preserve local/global RoPE behavior without changing ordinary text-only token prefill.
- Treat root text-only Gemma3 configurations as direct language-model configs unless they also declare a vision configuration, preserving existing text-session startup while retaining nested VLM defaults.
- Route CLI and continuous-batch image requests through a filtered host embedding table and resident IREE projector, with no duplicate decoder and no silent host/LLaVA fallback. Preserve the merged Qwen2-VL route and reject unqualified wrapper configs conservatively.
- Preserve Gemma3Processor's exact newline, BOI, image-token, EOI, newline block for both existing placeholders and bare CLI/server prompts, with checked capacity accounting for every wrapper token.
- Add a pinned mixed-runtime boundary diagnostic that captures processor pixels, twelve block-0 sub-stages, selected/final SigLIP states, post-layernorm, average-pool projection, prepared image rows, one-time `sqrt(hidden)` scaling, and the exact bidirectional padding mask.
- Expose that diagnostic as the standalone `xla_gemma3_reference_check` CUDA example so it avoids the libtest harness, while retaining the ignored test as a thin wrapper and emitting flushed stage progress plus 60-second heartbeats.
- Configure IREE's diagnostics-only process-global task threading exactly once before the standalone runner creates its first `local-task` instance: one topology group and host-default pthread stacks, with cached success/failure and no shared production `xla_aux`/`xla_iree` startup change.
- Reuse the existing eager and IREE hidden-state captures at exact ordered block 6, 13, 20, and final-block descriptors, partitioning blocks 1 through 26 into four balanced first-failure intervals without changing the graph, tolerances, or production path.
- Refine the known block-1-through-6 interval by comparing block outputs 1 through 5 before the existing block-6 checkpoint and emitting the exact first failing block while preserving the later bisection report.

## Validation

- [x] `cargo test -p mlxcel-xla --lib gemma3_` (8 passed)
- [x] Gemma3 diagnostics comparator test (1 passed)
- [x] `cargo test -p mlxcel-xla prepared_gemma3 --lib` (7 passed, including stable contract field sensitivity and unchanged text-only identity)
- [x] `cargo test -p mlxcel-xla --features iree gemma3_vlm_compatibility --lib` (2 passed: wrapper contract binding and direct-root text-only identity)
- [x] `cargo test -p mlxcel-xla gemma3_avg_pool_projector_binds_pool_and_prompt_contract --lib` (1 passed)
- [x] `cargo test -p mlxcel-xla qwen2_vl --lib` after rebasing #865 (4 passed)
- [x] `cargo test --features xla-iree --lib host_preprocessor` after rebasing #865 (13 passed)
- [x] `cargo test --lib vlm_prompt` (16 passed, including bare Gemma3 wrapping and unchanged unwrapped LLaVA insertion)
- [x] `cargo test -p mlxcel-xla gemma3 --lib` (52 passed, 2 ignored, including direct-root and nested-wrapper configuration regressions)
- [x] unqualified `llama`, `qwen2_vl`, and `gemma3` capability regression gate
- [x] `cargo check --features xla-iree --lib --bins`
- [x] `cargo fmt --all -- --check`
- [x] `IREE_DIST=... cargo check --no-default-features --features xla-reference-diagnostics --lib`
- [x] `IREE_DIST=... cargo clippy --no-default-features --features xla-reference-diagnostics --lib` (existing warnings only)
- [x] Cargo metadata confirms `xla_gemma3_reference_check` requires both `cuda` and `xla-reference-diagnostics`.
- [x] targeted Clippy for the corrected Gemma3 graph, diagnostics, and runtime compatibility contract (existing warnings only)
- [x] Real CUDA IREE projector execution using `/home/inureyes/models/gemma-3-4b-it-4bit`: compiled and returned finite `[256, 2560]` projected features in 32.17 seconds.
- [x] Post-fix production CUDA/IREE `mlxcel generate` run with the pinned checkpoint and image fixture completed successfully (`k`, 1 token, exit 0; reported generation time 8.11s).
- [x] Independent deterministic HF eager versus IREE `local-task` Gemma3 VLM embeddings/mask oracle: logits max absolute difference 0.006675, greedy token exact (`14`), all-layer K/V within max 0.25, RMS 0.1, and cosine 0.99 envelopes; causal-mask, multiplicative-mask, double-scale, and missing-image-pre-scale negatives all diverged.
- [x] Focused diagnostics-only threading regressions cover configuration-before-creation, error short-circuiting, exact two-flag source contract, native repeated-call reuse, and production `xla_aux`/`xla_iree` isolation (5 passed across filtered runs).
- [x] The linked diagnostics libtest contains both `xla_diagnostics_configure_local_task_threads` and `iree_flags_parse`; the native reuse test proved the pinned 3.11 parser accepts and consumes both synthetic flags.
- [x] `IREE_DIST=... cargo check --no-default-features --features cuda,xla-reference-diagnostics --example xla_gemma3_reference_check`
- [x] `IREE_DIST=... cargo test --features xla-reference-diagnostics --lib reference_boundary::` (10 passed, 1 actual ignored; includes exact balanced descriptors, ordered block-1-through-6 refinement descriptors, interval bounds, and exact-failure selection)
- [ ] The bounded standalone actual at `c44147f8` kept processor pixels, `siglip.hidden.embedding`, and every block-0 sub-stage passing. `siglip.hidden.block6.output` failed first with 333 failures, max absolute difference `1.760086`, and first failure at flat index 42197; block 13 failed with 35,808 failures and max absolute difference `1.969528`, block 20 failed with 489,930 failures and max absolute difference `68.06642`, and `siglip.hidden.last_pre_layernorm` failed with 1,408,110 failures and max absolute difference `1302.648`. The coarse report resolved `first_failing_block_range=1..=6`.
- [ ] The bounded standalone actual at `002bb77e` kept the prepared-prefill self/invariant and mask checks passing exactly. `siglip.hidden.block1.output` failed first with 8 failing values, max absolute difference `0.6672516`, and first failure at flat index 658517; block 2 failed with 18 values and max absolute difference `0.7214432`, block 3 with 61 and `1.209267`, block 4 with 73 and `1.210052`, block 5 with 133 and `1.209457`, and block 6 with 333 and `1.760086`.
- [ ] The ordered report resolved `exact_first_failing_block=1`. Exit 101 is the diagnostic's intended first-divergence panic at `siglip.hidden.block1.output` flat index 658517, not a worker-creation regression; numeric acceptance remains open. No further actual was run, and graph outputs, production behavior, tolerances, CSV handling, and diagnostic stack flags are unchanged.
- [ ] Padding/multi-image continuous batching, cancellation, and slot reuse remain required before this draft is ready.

## Runtime notes

- Gemma3 image execution requires the `xla-iree` feature and rejects `MLXCEL_XLA_VISION_BACKEND=host` explicitly.
- The standalone mixed-runtime runner requires MLX CUDA at build time and pins only the IREE side to `local-task`; immediately before its first IREE instance it applies `--task_topology_group_count=1` and `--task_worker_stack_size=0` once through the diagnostics-only flag parser.
- Text-only Gemma3 stays on the ordinary token-prefill path and retains its pre-existing dense artifact identity.
- Runtime artifacts are cached outside the repository; no generated CSV or model artifact is committed.

Closes #869

### CODE REVIEW

**Comment by inureyes** [comment]
Rebased onto `main@7fe1412d` and force-pushed as `c4e5d22d`.

Post-rebase validation:
- `cargo fmt --all -- --check`
- `git diff --check`
- `cargo test -p mlxcel-xla --lib gemma3_`: 8 passed
- `cargo check -p mlxcel-xla --features iree`: passed with existing warnings

The shared vision runtime inherited #934’s explicit `new_legacy_unqualified` artifact path during the rebase; no additional branch-local adaptation commit was needed. The strict SigLIP block-1 numeric failure (`8` values, prior `max_abs=0.6672516`) and continuous-batching gates remain unresolved. With CUDA unavailable on the current host, no fresh actual-checkpoint result is claimed. This PR remains draft and #869 remains blocked by #932.

**Comment by inureyes** [comment]
## Rebase and bounded SigLIP evidence — 2026-07-27

- Rebased the feature branch onto `origin/main@ec6fbb6d` and force-with-lease pushed `9ab13992`. The vision conflict adopts the shared `numeric_ops` LayerNorm/GELU/softmax helpers from main, the feature conflict preserves the CPU-capable `xla-micro-oracle`, and the diagnostics conflict adopts the shared one-group plus host-default-stack helper from main.
- The only authorized actual command was `env IREE_CUDA_HOME=/home/inureyes/.cache/mlxcel/iree-cuda-3.12.0rc20260721 IREE_CUDA_COMPILE=/home/inureyes/.cache/mlxcel/iree-cuda-3.12.0rc20260721/venv/bin/iree-compile MLXCEL_GEMMA3_REVISION=93724907d4ed1745d2fe50baadf3b0b01a65abf2 RUSTC=/home/inureyes/.rustup/toolchains/1.93.1-aarch64-unknown-linux-gnu/bin/rustc /home/inureyes/.rustup/toolchains/1.93.1-aarch64-unknown-linux-gnu/bin/cargo run --locked --example xla_gemma3_reference_check --features cuda,xla-reference-diagnostics -- --model /home/inureyes/models/gemma-3-4b-it-4bit --image tests/fixtures/test_image.png --device local-task`.
- That current-head actual was stopped at the bounded eight-minute cap before the model runner began. The new Rust 1.93.1 target caused `mlxcel-core` to classify its MLX cache as stale, purge `_deps`, and attempt the pinned MLX clone; the sandboxed attempt could not resolve GitHub. After allowing the required fetch, rebuilding pinned MLX revision `b7c3dd6d27f45b5365b08a840310187dc503f1db` completed and the same current-head example passed `cargo check` in 7m07s. No second actual or full-model run was started.
- Therefore there is no new current-head numeric result. The last completed actual remains the documented `002bb77e` run: first divergence `siglip.hidden.block1.output`, flat index `658517`, 8 failing values, max absolute difference `0.6672516`; exact first failing block `1`. This remains a numeric acceptance blocker and was not waived.
- Focused validation after the rebase: `cargo check --locked --features cuda,xla-reference-diagnostics --example xla_gemma3_reference_check` passed with existing warnings; `cargo test --locked -p mlxcel-xla --lib gemma3` passed 54 with 2 ignored; `cargo test --locked -p mlxcel-xla --lib diagnostic_flags::tests` passed 2; `cargo fmt --all -- --check` passed.
- No tolerance, production graph, CSV, or additional numeric fix was introduced. The PR remains draft; SigLIP intermediate acceptance, final token-exact output, padding/multi-image continuous batching, cancellation, and slot reuse remain open.

**Comment by inureyes** [comment]
## TF32 reference policy checked here: immaterial, so the numbers above stand

I filed on #932 that MLX defaults `MLX_ENABLE_TF32=1`, selecting a 10-bit-mantissa
`CUBLAS_COMPUTE_32F_FAST_TF32`, while `src/lib/mlxcel-xla/README.md` declares that mode is
not valid F32 reference evidence. That defect decided the Molmo2 gate in #916 (failed at the
first ViT block under the default, passes the whole gate with it off, 300x at `vit.block.0`),
so it had to be re-checked here before relying on the numbers above.

Re-run at `9ab13992` with `MLX_ENABLE_TF32=0`, everything else identical
(`local-task`, `MLX_CUDA_ARCHITECTURES=121`, unchanged `atol=0.08, rtol=0.04`). The result is
unchanged at every stage:

| stage | TF32 default | `MLX_ENABLE_TF32=0` |
| --- | --- | --- |
| `siglip.hidden.embedding` | 4.852448e-1, 0 fail | 4.852448e-1, 0 fail |
| `siglip.block0.output` | 3.646393e-1, 0 fail | 3.646393e-1, 0 fail |
| `siglip.hidden.block1.output` | 6.672516e-1, 8 fail | 6.672516e-1, 8 fail |
| `siglip.hidden.block6.output` | 1.760086, 333 fail | 1.760086, 333 fail |
| `siglip.hidden.block20.output` | 6.806642e1, 489930 fail | 6.806642e1, 489930 fail |
| `siglip.hidden.last_pre_layernorm` | 1.302648e3, 1408110 fail | 1.302648e3, 1408110 fail |

`exact_first_failing_block=1` and `first_failure=Some(658517)` are identical under both.

This is consistent with the mechanism: the flag only reaches genuine f32 contractions, and
this checkpoint's vision tower weights are BF16, so the SigLIP path never takes the TF32
route. The same check on the Qwen2.5-VL tower in #920 was likewise unchanged, while the two
Molmo gates, which do have f32 paths, both moved.

So unlike #916, the divergence recorded here is a property of the emitted graph rather than
of the reference policy, and the blocker in the PR body is unchanged.

For completeness, this run also reconfirms at current head the numbers the PR body carried
from `002bb77e`: block1 8 failures at `max_abs=0.6672516` first failure 658517, block2 18,
block3 61, block4 73, block5 133, block6 333, block13 35808, block20 489930,
`last_pre_layernorm` 1408110 at `max_abs=1302.648`. Every block-0 substage passes
(`layer_norm1`, `q_proj`, `k_proj`, `v_proj`, `attention_context`, `attention_output`,
`attention_residual`, `layer_norm2`, `mlp_fc1`, `mlp_activation`, `mlp_fc2`, `output`), as do
the processor pixels, the prepared text/image identity and both mask cell checks, all exact.
So the stale-evidence concern for this PR is closed: the recorded numbers are current-head
numbers.
